### PR TITLE
[WIP] Interface implementation generator (fixes #1568)

### LINKED ISSUE #1568: Support 'Generate' actions
I'd like to have a "Generate" action that enables me to implement all the methods defined in a interface on a certain struct, just like I can use it in Java to add the boilerplate code to implement a certain method / interface / abstract class.

The way it should work would be like this:
Allow you to search in the list of all existing interfaces, once you've selected which interfaces you want to use generate the method stubs for those interfaces bound to the struct and add a new "constructor" as defined below.

Ideally there should be two "Generate" actions, one to generate an implementation on a pointer struct and one on the concrete struct.

Concrete example:

```go
package awesome

type (
	Fetcher interface {
		Execute(string) string
	}

	myFetcher struct <caret> {
		myDependency string
	}
)

```

Expected result:

```go
package awesome

type (
	Fetcher interface {
		Execute(string) string
	}

	myFetcher struct {
		myDependency string
	}
)

func (mf *myFetcher) Execute(<caret> string) string {
	//TODO Add implementation
}


func NewFetcher() Fetcher {
	return &myFetcher {}
}
```

### PR DESCRIPTION
**Pull request by dlsniper**
This is a draft implementation for the generator implementation. It's missing the ability to search for existing implementations and a way to add the generated method implementations to the code.

### CODE REVIEW

**Comment by zolotov** [comment]
Generate action is something that generates the code in place of caret. You're applying action on specific struct element, I believe intention action is more appropriate.
screenshot 2015-04-27 13 36 12

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                                                      final boolean withPointerReference) {
+    if (interfaces.size() == 0) {
+      return;
+    }
+
+    new WriteCommandAction.Simple(struct.getProject(), struct.getContainingFile()) {
+      @Override
+      protected void run() throws Throwable {
+
+        StringBuilder prototypeBuffer = new StringBuilder("func (");
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: "func (func ("? I didn't get it.

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        prototypeBuffer.append("func (")
+          .append(struct.getStruct().getText().substring(0, 1).toLowerCase());
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: So it's always 's', right?

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if (withPointerReference) {
+          prototypeBuffer.append("*");
+        }
+        prototypeBuffer.append(struct.getStruct().getText()).append(") ");
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: And it always 'struct'

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        StringBuilder builder = new StringBuilder("");
+        for (GoInterfaceType goInterface : interfaces) {
+          for (GoMethodSpec method : goInterface.getMethods()) {
+            builder.append(prototype)
+              .append(method.getSignature())
+              .append(" {\n\t// TODO implement this\n}\n");
+          }
+        }
+
+        builder.toString();
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: What is this?

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  private static void generateInterfaceImplementation(@NotNull final GoStructType struct,
+                                                      @NotNull final List<GoInterfaceType> interfaces,
+                                                      final boolean withPointerReference) {
+    if (interfaces.size() == 0) {
+      return;
+    }
+
+    new WriteCommandAction.Simple(struct.getProject(), struct.getContainingFile()) {
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: In the end you don't run the action. Btw there is a shortcut `WriteCommandAction.runWriteCommandAction`

File: src/com/goide/generate/GenerateInterfaceImplementationAction.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  private static GoStructType getPsiStructFromContext(@NotNull AnActionEvent e) {
+    PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
+    Editor editor = e.getData(CommonDataKeys.EDITOR);
+    if (psiFile == null || editor == null) {
+      return null;
+    }
+
+    int offset = editor.getCaretModel().getOffset();
+    PsiElement elementAt = psiFile.findElementAt(offset);
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationAction.java]: CommonDataKeys#PSI_ELEMENT

File: src/com/goide/generate/GenerateInterfaceImplementationDialog.java [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+
+    JBTextField searchBox = new JBTextField();
+    searchBox.addActionListener(getSearchListener());
+    myPanel.add(LabeledComponent.create(searchBox, "Interface name"));
+
+    // TODO better interface here
+    // We should have a search box and when a user types something in
+    // completion should kick in and present the user a list of interfaces that match
+    // After the desired interface is found, it should be added to the list
+    mySelectedInterfaces = new CollectionListModel<GoInterfaceType>(new GoInterfaceType() {
```
Comment (zolotov) [inline src/com/goide/generate/GenerateInterfaceImplementationDialog.java]: I just don't understand this.

**Comment by dlsniper** [comment]
I'll close this for now so either it can be picked up by someone else, or I'll do it properly when I get some time if not.

**Pull request closed without merging**